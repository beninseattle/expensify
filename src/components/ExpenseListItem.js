import React from 'react';
import {Link} from 'react-router-dom';
import paths from '../routers/paths';

/**
 * @param {object} props
 * @param {number} props.index
 * @param {Expense} props.expense
 * @constructor
 */
const ExpenseListItem = ({index, expense}) => (
  <Link className="list-item" to={paths.editExpense + expense.id}>
    <div>
      <h3 className="list-item__title">{expense.description}</h3>
      <span className="list-item__subtitle">({expense.createdAtDateString()})</span>
    </div>
    <h3 className="list-item__data">{expense.amountCurrencyString()}</h3>
  </Link>
);

export default ExpenseListItem;