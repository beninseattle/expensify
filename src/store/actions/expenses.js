import database from '../../firebase/firebase';
import Expense from '../../models/expense';

/**
 * @callback expenseAction
 * @param {Expense} expense
 */
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

/**
 * Returns function to start async action to add a new expense in Firebase and then dispatch to redux
 * @param {Expense} expense
 * @returns {function(func,func)}
 */
export const startAddExpense = (expense) => {
  // function functionality thanks to redux-thunk
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses`).push(expense.dataForSave())
      .then((ref) => {
        expense.saveToStore(ref.key);
        dispatch(addExpense(expense));
      })
      .catch((error) => {
        console.log(`Add expense for expense ${expense.id} failed: ${error}`);
      });
  };
};

/**
 * @callback expenseAction
 * @param {Expense} expense - the expense to remove from the store
 */
export const deleteExpense = (expense) => ({
  type: 'DELETE_EXPENSE',
  expense
});

/**
 * Returns function to start async action to delete a specific expense in Firebase and then dispatch to redux
 * @param expense
 * @returns {function(func, func)}
 */
export const startDeleteExpense = (expense) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${expense.id}`).remove()
      .then(() => {
        dispatch(deleteExpense(expense));
      })
      .catch((error) => {
        console.log(`Delete for expense ${expense.id} failed: ${error}`);
      });
  };
};

/**
 * @callback expenseAction
 * @param {Expense} expense - object with 1 or more of the fields of an expense
 */
export const editExpense = (expense) => ({
  type: 'EDIT_EXPENSE',
  expense
});

/**
 * Returns function to start async action to save a specific expense in Firebase and then dispatch to redux
 * @param expense
 * @returns {function(func, func)}
 */
export const startEditExpense = (expense) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${expense.id}`).set(expense.dataForSave())
      .then(() => {
        dispatch(editExpense(expense));
      })
      .catch((error) => {
        console.log(`Edit for expense ${expense.id} failed: ${error}`);
      });
  };
};

/**
 * @callback setExpenses
 */
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

/**
 * Returns function to start async action to load expenses from Firebase and then dispatch to redux
 * @returns {function(func, func)}
 */
export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses`).once('value')
      .then((snapshot) => {
        const expenses = [];
        snapshot.forEach((childSnapshot) => {
          expenses.push(new Expense({
            id: childSnapshot.key,
            ...childSnapshot.val()
          }));
        });

        dispatch(setExpenses(expenses));
      })
      .catch((error) => {
        console.log(`Set expenses failed to load data: ${error}`);
      });
  };
};