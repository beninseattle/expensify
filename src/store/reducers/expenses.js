const expensesReducerDefaultState = [];
export const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense];

    case 'DELETE_EXPENSE':
      return state.filter(({id}) => id !== action.expense.id);

    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.equals(action.expense.id)) {
          return action.expense;
        } else {
          return expense;
        }
      });

    case 'SET_EXPENSES':
      return action.expenses;

    default:
      return state;
  }
};
