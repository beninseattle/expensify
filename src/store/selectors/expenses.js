import moment from 'moment';

/**
 * Filtering by dates compares based on a granularity of days
 *
 * @param expenses
 * @param filters
 */
export default (expenses, filters) => {
  let visibleExpenses = expenses.filter((expense) => {
    if (filters.text.length > 0) {
      if (expense.note.toLowerCase().indexOf(filters.text.toLowerCase()) === -1 &&
        expense.description.toLowerCase().indexOf(filters.text.toLowerCase()) === -1) {
        return false;
      }
    }
    if (filters.startDate !== undefined) {
      if (moment(expense.createdAt).isBefore(filters.startDate, 'day')) {
        return false;
      }
    }
    if (filters.stopDate !== undefined) {
      if (moment(expense.createdAt).isAfter(filters.stopDate, 'day')) {
        return false;
      }
    }
    return true;
  });

  visibleExpenses.sort((a, b) => {
    if (filters.sortBy === 'date') {
      if (a.createdAt === b.createdAt) {
        return 0;
      } else {
        return a.createdAt < b.createdAt ? -1 : 1;
      }
    } else if (filters.sortBy === 'amount') {
      if (a.amount === b.amount) {
        return 0;
      } else {
        return a.amount < b.amount ? -1 : 1;
      }
    }
    return 0;
  });
  return visibleExpenses;
};
