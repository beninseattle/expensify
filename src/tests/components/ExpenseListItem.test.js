import React from 'react';
import {shallow} from 'enzyme';
import ExpenseListItem from "../../components/ExpenseListItem";
import expenses from '../fixtures/expenses';

test('should render ExpenseListItem with an expense', () => {
  const wrapper = shallow(<ExpenseListItem index={0} expense={expenses[0]}/>);

  expect(wrapper).toMatchSnapshot();
});
