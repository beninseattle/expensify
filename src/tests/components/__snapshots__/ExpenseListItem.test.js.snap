// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with an expense 1`] = `
<div>
  <h3>
    <Link
      replace={false}
      to="/edit/1"
    >
      abc
    </Link>
  </h3>
  <p>
    US$Â 10.00
     (
    Saturday, Dec 31, 2016
    )
  </p>
  <withRouter(theButton)
    path="/edit/1"
    title="Edit"
  />
</div>
`;
